sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs

```diff
@@ -282,57 +282,82 @@ private static void CloseLink(RequestResponseAmqpLink link)
             {
                 if (!_receiveLink.TryGetOpenedObject(out link))
                 {
-                    link = await _receiveLink.GetOrCreateAsync(UseMinimum(_connectionScope.SessionTimeout, timeout)).ConfigureAwait(false);
+                    link = await _receiveLink.GetOrCreateAsync(UseMinimum(_connectionScope.SessionTimeout, timeout))
+                        .ConfigureAwait(false);
                 }
+
                 cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
-                var messagesReceived = await Task.Factory
-                    .FromAsync<(ReceivingAmqpLink, int, TimeSpan?, TimeSpan), IEnumerable<AmqpMessage>>
-                (
-                    static (arguments, callback, state) =>
-                    {
-                        var (link, maxMessages, maxWaitTime, timeout) = arguments;
-                        return link.BeginReceiveRemoteMessages(
-                            maxMessages,
-                            TimeSpan.FromMilliseconds(20),
-                            maxWaitTime ?? timeout,
-                            callback,
-                            link);
-                    },
-                    static asyncResult =>
-                    {
-                        var link = (ReceivingAmqpLink)asyncResult.AsyncState;
-                        bool received = link.EndReceiveMessages(asyncResult, out IEnumerable<AmqpMessage> amqpMessages);
-                        return received ? amqpMessages : Enumerable.Empty<AmqpMessage>();
-                    },
-                    (link, maxMessages, maxWaitTime, timeout),
-                    default
+                var receiveMessagesCompletionSource =
+                    new TaskCompletionSource<IEnumerable<AmqpMessage>>(TaskCreationOptions
+                        .RunContinuationsAsynchronously);
+
+                using var registration = cancellationToken.Register(static state =>
+                {
+                    var tcs = (TaskCompletionSource<IEnumerable<AmqpMessage>>)state;
+                    tcs.TrySetCanceled();
+                }, receiveMessagesCompletionSource, useSynchronizationContext: false);
+
+                // in case BeginReceiveRemoteMessages throws exception will be materialized on the synchronous path
+                _ = Task.Factory
```

Comment:
At first I had an approach following this pattern

```
var receiveTask = Receive(...)
var completed = await Task.WhenAny(receiveTask, tcs.Task).ConfigureAwait(false)
if(completed == tcs.Task)
   await tcs.Task.ConfigureAwait()

await receiveTask.ConfigureAwait()
```

Reply:
But then I figure with this approach we always pay the price of the array allocation of `WhenAny` plus the additional conditions on the path including the state machinery. So I ended up always using the TCS to materialize either the result or the cancellation.

Reply:
Let me know if you prefer the `WhenAny` approach. While it has more state machine involved and allocates the WhenAny array it wouldn't require us to box the value tuple and might make the code slightly more straightforward to read at the cost of more allocations

Reply:
I like your implementation better;  I sketched out the `WhenAny` approach in the issue, but had no idea that `Register` was a thing.   While this does introduce a bit more density in the `FromAsync` machinery, it already had some complexity,   With your current approach, I find the flow easier to follow than the "which task completed" juggling that `WaitAny` would require.